### PR TITLE
HIVE-28207: NullPointerException is thrown when checking column uniqueness

### PR DESCRIPTION
**Pull request by okumin**
### What changes were proposed in this pull request?

Add a null check when using `RelMetadataQuery#areColumnsUnique`.
https://issues.apache.org/jira/browse/HIVE-28207

### Why are the changes needed?

`RelMetadataQuery#areColumnsUnique` returns `true,` `false,` or `null.` Some of our implementations skip testing whether it is null and throw NPE.

### Does this PR introduce _any_ user-facing change?

No. It is a bug fix to resolve NPE.

### Is the change a dependency upgrade?

No.

### How was this patch tested?

I added integration tests.
Without this patch, `cbo_join_constraints.q` fails with NPE thrown by `HiveJoinConstraintsRule`.
`cbo_row_count_non_simple_filter.q` hits the problem of `HiveJoinConstraintsRule` and `HiveRelMdRowCount`.
I have not found a real case where the issue happens with `HiveAggregateSplitRule`, but I modified it as it is likely to hit the same problem.

### CODE REVIEW

**Comment by ayushtkn** [comment]
> RelMetadataQuery#areColumnsUnique returns true, false, or null.

why is it returning `null`, columns can be either unique or not unique, what `null` denotes here?

**Comment by okumin** [comment]
@ayushtkn Calcite can express the three statuses, either of "it is definitely unique", "it is not definitely unique", or unknown. Null stands for the last case. If it is null, this PR assumes a more general case, not unique.
https://github.com/apache/calcite/blob/0551b8903391c1706422a2c1b8b648a6941f39a2/core/src/main/java/org/apache/calcite/rel/metadata/RelMetadataQuery.java#L523-L524

**Comment by okumin** [comment]
I may also check why and when it can return null if it is unexpected

**Review by zabetak — APPROVED** [review]
Overall LGTM. I think it makes sense to add checks for nulls when necessary since from an API perspective it is a valid value. I left some minor comments in .q files but other than that I am fine to merge this change.

Independently (maybe as a separate JIRA ticket) it may be worth checking why the null value appears in the first place and tackle that problem as well since it could lead to better query plans. In many cases the NPE raised in metadata can be treated by finding the origin of the null value.

File: ql/src/test/queries/clientpositive/cbo_join_constraints.q [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+EXPLAIN CBO
+SELECT i_item_sk
+FROM store_sales, household_demographics, item
+WHERE ss_item_sk = i_item_sk;
+
+EXPLAIN
+SELECT i_item_sk
+FROM store_sales, household_demographics, item
+WHERE ss_item_sk = i_item_sk;
```
[the remark this answers is not in the document]
Comment (okumin) [reply ql/src/test/queries/clientpositive/cbo_join_constraints.q]: Thanks for the suggestion. Updated.
https://github.com/apache/hive/pull/5207/commits/e3abdf3a1018620db175f2678d59e9536cbf062c

**Review by ayushtkn — APPROVED** [review]
LGTM

**Comment by okumin** [comment]
CI succeeded 💯

**Pull request closed without merging**